src/random_bq_to_anndata.py

```diff
@@ -1,46 +1,178 @@
     query = client.query(sql)
     return query.result()
 
 
-def random_bq_to_anndata(project, dataset, num_cells):
+def random_bq_to_anndata(project, dataset, num_cells, output_file_prefix):
     client = bigquery.Client(project=project)
-    cell_data = get_cell_data(project, dataset, client, num_cells)
-    adata = ad.AnnData()
-    for row in list(cell_data):
-        # insert VS-507 here
-        pass
+
+    print(f"Getting {num_cells} random cells' data from {project}.{dataset}...")
+    random_cell_ids = get_random_cell_ids(project, dataset, client, num_cells)
+
+    # Read the cell information and store metadata for later
+    ordered_cell_indices, cell_index_to_cell = get_cells(project, dataset, client, random_cell_ids)
+
+    original_cell_ids = []
+    cell_types = []
+    for cell_index in ordered_cell_indices:
+        original_cell_ids.append(cell_index_to_cell[cell_index].original_cell_id)
+        cell_types.append(cell_index_to_cell[cell_index].cell_type)
+
+    # Read the feature information and store metadata for later.
+    ordered_feature_indices, feature_index_to_feature = get_features(project, dataset, client)
+
+    feature_ids = []
+    feature_names = []
+    last_feature_index = None
+    for feature_index in ordered_feature_indices:
+        if (last_feature_index is None):
+            last_feature_index = feature_index - 1
+        # # We expect all entries in cas_feature_info to be stored in continuous fashion. (1, 2, 3, ...)
+        if (feature_index != last_feature_index + 1):
+            raise Exception("ERROR: Non-continuous values for `cas_feature_index` in table `cas_feature_info`")
+
+        feature_ids.append(feature_index_to_feature[feature_index].original_feature_id)
+        feature_names.append(feature_index_to_feature[feature_index].feature_name)
+        last_feature_index = feature_index
+
+    # Note that this method requires that the result set returned by get_cell_data be sorted by cas_cell_index (or, could also be sorted by original_cell_id)
+    cell_data = get_matrix_data(project, dataset, client, random_cell_ids)
+
+    (index_ptr, indices, data) = generate_sparse_matrix(cell_data, ordered_feature_indices[0])
+
+    # Create the matrix from the sparse data representation generated above.
+    counts = csr_matrix((data, indices, index_ptr), dtype=np.float32)
+    adata = ad.AnnData(counts)
+    adata.obs_names = original_cell_ids
+    adata.obs["cell_type"] = cell_types
+    adata.var_names = feature_ids
+    adata.var["feature_name"] = feature_names
+
+    # See https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.raw.html?highlight=raw#anndata.AnnData.raw
+    #  for why we set 'raw' thusly: "The raw attribute is initialized with the current content of an object by setting:"
+    adata.raw = adata
+    adata.write(f'{output_file_prefix}.h5ad', compression="gzip")
```

Comment:
PyCharm is not happy with this, it wants the first arg to be a `PathLike` rather than a string

Reply:
Okay, I've made the first arg a `Path`- don't have PyCharm running on my machine, so let me know if it's still not happy.